src/kaleidoscope_internal/LEDModeManager.h

```diff
@@ -0,0 +1,428 @@
+/* Kaleidoscope - Firmware for computer input devices
+ * Copyright (C) 2013-2018  Keyboard.io, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "kaleidoscope_internal/array_like_storage.h"
+#include "kaleidoscope/plugin.h"
+#include "kaleidoscope/plugin/LEDMode.h"
+
+#include <stddef.h>
+
+// To enable placement new, we need to supply a global operator
+// function.
+//
+inline void* operator new (size_t, void* __p) throw() {
+  return __p;
+}
+
+namespace kaleidoscope {
+
+namespace plugin {
+// Forward declarations to avoid header inclusions
+class LEDControl;
+class LEDModeInterface;
+} // end namespace plugin
+
+namespace internal {
+namespace led_mode_management {
+
+// Functions of this type allocate a new LED mode within a global
+// buffer and give it access to it's parent plugin.
+// This allows a dynamic LED mode to access member data of its parent
+// plugin, which can help to reduce its RAM footprint.
+// Some dynamic LED modes might ignore the parent pointer entirely, e.g. if
+// they do not have a configurable internal state.
+//
+// We use a factory function instead of a polymorphic
+// class to generate LED modes. By this means we can avoid the additional
+// cost of vtables which safes some program memory.
+// Conceptually, the factory function approach is similar to having
+// a polymorphic class with only one method, that is accessed directly
+// instead of the detour via the vtable.
+//
+typedef kaleidoscope::plugin::LEDMode*
+(*LEDModeFactoryFunc)(void *raw_buffer,
+                      void *parent,
+                      uint8_t mode_id
+                     );
```

Comment:
Can you explain why the `typedef` is here?

Reply:
Because function pointer types are ugly to read, so you only want to write them once and then go with the typedef-ed type.

Reply:
In that case, I guess I have a more fundamental question – what does this line do? It looks like a combination of a definition of a type alias and a function declaration, and doesn't make any sense to me at all.

Reply:
That defines a typename alias for a function pointer. The syntax is different from that of ordinary (non-function pointer) typedefs. In general
```cpp
typedef ReturnType (*PtrTypeAlias)(...arguments...);
```
says that `PtrTypeAlias` can furtheron be the type of a pointer that points to a function of given type.

Example
```
typedef int (*f)(int); // function pointer alias
int f1(int a) { return a; } // function of same type

int main() {
    f my_f_1 = f1; // function pointer assignment
    f my_f_2 = &f1; // alternative syntax
    (*my_f_1)(10); // function invocation via function pointer
}
```